File: Payload_Type/poseidon/mythic/agent_functions/curl.py
from mythic_payloadtype_container.MythicCommandBase import *
import base64


class CurlArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="url",
                type=ParameterType.String,
                description="URL to request.",
                default_value="https://www.google.com",
                parameter_group_info=[
                    ParameterGroupInfo(ui_position=1)
                ]
            ),
            CommandParameter(
                name="method",
                type=ParameterType.ChooseOne,
                description="Type of request",
                choices=["GET", "POST"],
                parameter_group_info=[
                    ParameterGroupInfo(ui_position=2)
                ]
            ),
            CommandParameter(
                name="headers",
                type=ParameterType.String,
                description="JSON string of headers: {\"Host\":\"a.b.c.com\"}",
                parameter_group_info=[
                    ParameterGroupInfo(ui_position=3, required=False)
                ]
            ),
            CommandParameter(
                name="body",
                type=ParameterType.String,
                description="Body contents to send in request",
                parameter_group_info=[
                    ParameterGroupInfo(ui_position=4, required=False)
                ]
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) == 0:
            raise Exception("Must provide arguments")
        else:
            try:
                self.load_args_from_json_string(self.command_line)
                self.add_arg("method", self.get_arg("method").upper())
                if self.get_arg("headers") is not None:
                    self.add_arg("headers", base64.b64encode(self.get_arg("headers").encode()).decode())
                else:
                    self.add_arg("headers", "")
                if self.get_arg("body") is not None:
                    self.add_arg("body", base64.b64encode(self.get_arg("body").encode()).decode())
                else:
                    self.add_arg("body", "")
            except:
                raise Exception("Failed to process arguments as JSON.")

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)
        self.add_arg("method", self.get_arg("method").upper())
        if self.get_arg("headers") is not None:
            self.add_arg("headers", base64.b64encode(self.get_arg("headers").encode()).decode())
        else:
            self.add_arg("headers", "")
        if self.get_arg("body") is not None:
            self.add_arg("body", base64.b64encode(self.get_arg("body").encode()).decode())
        else:
            self.add_arg("body", "")


class CurlCommand(CommandBase):
    cmd = "curl"
    needs_admin = False
    help_cmd = 'curl -url https://www.google.com -method GET'
    description = "Execute a single web request."
    version = 1
    author = "@xorrior"
    argument_class = CurlArguments
    attackmapping = ["T1213"]

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        task.display_params = task.args.get_arg("url") + " via HTTP " + task.args.get_arg("method")
        if len(task.args.get_arg("headers")) > 0:
            task.display_params += " with custom headers"
        if len(task.args.get_arg("body")) > 0:
            task.display_params += " with custom body"
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/curl.go
package agentfunctions

import (
	"encoding/base64"
	"errors"
	"fmt"

	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "curl",
		Description:         "Execute a single web request",
		HelpString:          "curl -url https://www.google.com -method GET -headers \"Host: abc.com\" -headers \"Authorization: Bearer $TOKEN\"",
		Version:             1,
		Author:              "@xorrior",
		MitreAttackMappings: []string{"T1213"},
		SupportedUIFeatures: []string{},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{},
		},
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:          "url",
				DefaultValue:  "https://www.google.com",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
					},
				},
				Description: "URL to request",
			},
			{
				Name:             "method",
				ModalDisplayName: "HTTP Method",
				DefaultValue:     "GET",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
				Choices:          []string{"GET", "POST", "PUT", "DELETE"},
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     2,
					},
				},
				Description: "Type of request",
			},
			{
				Name:             "headers",
				ModalDisplayName: "HTTP Headers",
				DefaultValue:     []string{},
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_ARRAY,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     3,
					},
				},
				Description: "Array of headers in Key: Value entries",
			},
			{
				Name:             "body",
				ModalDisplayName: "Body",
				DefaultValue:     "",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     4,
					},
				},
				Description: "Body contents to send in request",
			},
			{
				Name:             "socketPath",
				ModalDisplayName: "Unix Socket Path",
				CLIName:          "socketPath",
				DefaultValue:     "",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
					},
				},
				Description: "Path to UNIX Socket if you want to use that instead of a remote host",
			},
		},
		TaskFunctionCreateTasking: func(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  taskData.Task.ID,
			}
			url, err := taskData.Args.GetStringArg("url")
			if err != nil {
				logging.LogError(err, "Failed to get url string")
				response.Success = false
				response.Error = err.Error()
				return response
			}
			method, err := taskData.Args.GetStringArg("method")
			if err != nil {
				logging.LogError(err, "Failed to get method string")
				response.Success = false
				response.Error = err.Error()
				return response
			}
			bodyString, err := taskData.Args.GetStringArg("body")
			if err != nil {
				logging.LogError(err, "Failed to get body string")
				response.Success = false
				response.Error = err.Error()
				return response
			}
			taskData.Args.SetArgValue("body", base64.StdEncoding.EncodeToString([]byte(bodyString)))
			displayParams := fmt.Sprintf("%s via HTTP %s", url, method)
			socketPath, err := taskData.Args.GetStringArg("socketPath")
			if err != nil {
				logging.LogError(err, "Failed to get socketPath")
				response.Success = false
				response.Error = err.Error()
				return response
			}
			if socketPath != "" {
				displayParams += fmt.Sprintf(" to %s", socketPath)
			}
			response.DisplayParams = &displayParams
			return response
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			if len(input) > 0 {
				return args.LoadArgsFromJSONString(input)
			} else {
				return errors.New("Must supply arguments")
			}
		},
	})
}
